### ISSUE TITLE
OSv crash in a weird way when it receive permission denied from nfs_open()

### ISSUE THREAD

**Issue by benoit-canet** [cloudius-systems/osv#706]
Using:

./scripts/run.py -d  -V -v --novnc  -e '/tools/mount-nfs.so nfs://192.168.89.5/fff?uid=0 /fff; /cli/cli.so'

VFS: mounting nfs at /fff
/# cat /fff/blub
page fault outside application, addr: 0x0000000000000000
[registers]
RIP: 0x00000000006f5194 <???+7295380>
RFL: 0x0000000000010202  CS:  0x0000000000000008  SS:  0x0000000000000010
RAX: 0xffffa00004235f00  RBX: 0x0000000000000000  RCX: 0x00002000002fd620  RDX: 0x00000000006fe830
RSI: 0x0000000000000000  RDI: 0x0000000000000000  RBP: 0xffffa00004235f00  R8:  0xffffa00003ce5340
R9:  0x00002000002fd5b0  R10: 0x0000000000000036  R11: 0x0000000000000001  R12: 0x00002000002fd620
R13: 0x00002000002fd5b0  R14: 0x0000000000000036  R15: 0x0000000000000001  RSP: 0x00002000002fd5f0
Aborted

[backtrace]
0x000000000022d8f0 <abort(char const*, ...)+249>
0x00000000003c7fc1 <???+3964865>
0x00000000003c8160 <mmu::vm_fault(unsigned long, exception_frame*)+350>
0x000000000048d29d <page_fault+315>
0x000000000048c126 <???+4768038>
0xffff90000506efff <???+84340735>
0x00000000ffffffff <???+-1>

Here the server was missing the no_root_squash option on my test setup and nfs_open() returned
EACCESS because request where mapped to the anonymous uid/guid.

This is clearly a server configuration and the nfs_op_open() is return EACCES to the VFS layer
but OSv does not handle it well and crash in a cryptic ways

**Comment by benoit-canet** [comment]
@justinc1 Could you confirm that putting no_root_squash on your server setup resolve the issue
then close the list /mount_point bug and confirm this one ?

**Comment by justinc1** [comment]
Yes, with no_root_squash server-side and uid=0 client-side, I can read file from NFS share.

**Comment by justinc1** [comment]
An update. A working example - clients works - ls shows file, cat shows file content.
```
Server exports
/fff          172.16.0.0/16(rw,sync,no_subtree_check,insecure,no_root_squash)
/fff       192.168.122.0/24(rw,sync,no_subtree_check,insecure,no_root_squash)

client:
sudo ./scripts/run.py -d -V -n -v -e '/tools/mount-nfs.so nfs://192.168.122.1/fff/?uid=0 /fff; /cli/cli.so'
```

If client is run with uid=1000, ls shows file, but cat fails.

With combined nfsv4 and nfsv3 like exports, neither ls nor cat work:
```
# nfsv3
/home/justin_cinkelj/devel/mikelangelo 192.168.122.0/24(rw,sync,no_subtree_check,no_root_squash,insecure,sec=sys)
# nfsv4
/export  192.168.122.0/24(fsid=0,rw,sync,no_subtree_check,no_root_squash,insecure,crossmnt,sec=sys)
/export/home/justin_cinkelj/devel/mikelangelo 192.168.122.0/24(rw,sync,no_subtree_check,no_root_squash,insecure,sec=sys)

client
sudo ./scripts/run.py -d -V -n -v -e '/tools/mount-nfs.so nfs://192.168.122.1/home/justin_cinkelj/devel/mikelangelo/?uid=0 /fff; /cli/cli.so'

server log:
Jan 19 10:55:24 jcpc rpc.mountd[1405]: /home/justin_cinkelj/devel/mikelangelo and /export/home/justin_cinkelj/devel/mikelangelo have same filehandle for 192.168.122.0/24, using first
Jan 19 10:56:32 jcpc rpc.mountd[1405]: authenticated mount request from 192.168.122.210:54620 for /home/justin_cinkelj/devel/mikelangelo (/home/justin_cinkelj/devel/mikelangelo)

client
ls /fff
client crash
changing sec=sys to sec=None or to remove it doesn't change anything. 
```

**Comment by justinc1** [comment]
Crash occurs also with 'working' /fff - I just have to create a subdirectory inside /fff.

**Comment by benoit-canet** [comment]
On Tue, Jan 19, 2016 at 11:00 AM, justinc1 <notifications@github.com> wrote:

> An update. A working example - clients works - ls shows file, cat shows
> file content.
>
> Server exports
> /fff          172.16.0.0/16(rw,sync,no_subtree_check,insecure,no_root_squash)
> /fff       192.168.122.0/24(rw,sync,no_subtree_check,insecure,no_root_squash)
>
> client:
> sudo ./scripts/run.py -d -V -n -v -e '/tools/mount-nfs.so nfs://192.168.122.1/fff/?uid=0 /fff; /cli/cli.so'
>
> If client is run with uid=1000, ls shows file, but cat fails.
>

Looks like permission denied.


> With combined nfsv4 and nfsv3 like exports, neither ls nor cat work:
>
> # nfsv3
> /home/justin_cinkelj/devel/mikelangelo 192.168.122.0/24(rw,sync,no_subtree_check,no_root_squash,insecure,sec=sys)
> # nfsv4
> /export  192.168.122.0/24(fsid=0,rw,sync,no_subtree_check,no_root_squash,insecure,crossmnt,sec=sys)
> /export/home/justin_cinkelj/devel/mikelangelo <http://192.168.122.0/24(fsid=0,rw,sync,no_subtree_check,no_root_squash,insecure,crossmnt,sec=sys)/export/home/justin_cinkelj/devel/mikelangelo> 192.168.122.0/24(rw,sync,no_subtree_check,no_root_squash,insecure,sec=sys)
>
>
Looks like sec=sys has nothing to do with NFSv4 but everything to do with
permissions:

https://www.centos.org/docs/5/html/Deployment_Guide-en-US/s1-nfs-client-config-options.html
(search for sec=sys)

It's a client side option.

Informations about exporting NFS v4 can be found here:

https://www.centos.org/docs/5/html/Deployment_Guide-en-US/s1-nfs-server-config-exports.html

**Comment by justinc1** [comment]
> sec=sys
> It's a client side option.

Ah, thanks for info - I was guessing I might/must be doing some things wrong, so I posted parts of config too. Seems server just ignored invalid option (or got silently confused?) - I didn't notice any error/warning in my logs.

**Comment by benoit-canet** [comment]
I believe that NFS v3 and v4 are so different (no portmap or mountd) that
OSv should not even see a NFS v4 server.

**Comment by nyh** [comment]
@justinc1 , @benoit-canet , and you please confirm that the latest commits (specifically 46fcceadbe076e66f11136bf99f14085505fead6 and e3e49419c9dc3ee8151e70170c01381e4beaf94a) fixed the various variants of this issue (with cat, ls, etc.)?

**Comment by benoit-canet** [comment]
works for me with cat and ls.

**Comment by justinc1** [comment]
Glad to say - it works for me too :) Tried ls, cat, mkdir, fopen and write - all good. It works also when share is mounted with uid=1000 - on server, files are then owned by uid=1000,gid=0.

**Comment by nyh** [comment]
Excellent. I'm happy we ended up with finding and fixing the real bugs (one in the REST API and one in OSv itself), and not making workarounds. I'm closing this issue. Thanks @benoit-canet and @justinc1 !

**Comment by justinc1** [comment]
One additional path to failure found. HLRS testbed server exports is:
```
/storage/mikelangelo/ssd_scratch 172.18.2.0/24(rw,no_root_squash,async,insecure)
```

I try to mount with
```
    command2 = '/tools/mount-nfs.so nfs://172.18.2.3/storage/mikelangelo/ssd_scratch/justin/openfoam/?uid=0 /openfoam'

    command2 = '/tools/mount-nfs.so nfs://172.18.2.3/storage/mikelangelo/ssd_scratch/?uid=0 /scratch'
```

First one will crash OSv, second one works. So subdirectory of NFS export cannot be mounted. I can live with second case, but its worth to at least mention.

**Comment by benoit-canet** [comment]
I can reproduce the bug it does appear at random.

**Comment by benoit-canet** [comment]
@justinc1 The bug is random here could you paste the crash as a comment ?

**Comment by justinc1** [comment]
A very generic one, I think.
```
page fault outside application, addr: 0x0000000000000000
[registers]
RIP: 0x000000000058daf5 <???+5823221>
RFL: 0x0000000000010206  CS:  0x0000000000000008  SS:  0x0000000000000010
RAX: 0x0000000000000000  RBX: 0xffffa00007bd8b00  RCX: 0xffff800005c48050  RDX: 0xffff800005c48050
RSI: 0x000000000000ffff  RDI: 0x0000000000000000  RBP: 0x00002000006ff070  R8:  0xffffa00007b96c80
R9:  0xffffa000034c79e0  R10: 0x0000000000000073  R11: 0x0000000000000000  R12: 0x00000000006e6db0
R13: 0xffffa000034c79e0  R14: 0x0000000000000073  R15: 0x0000000000000000  RSP: 0x00002000006ff060
Aborted

[backtrace]
0x000000000022bc6d <abort(char const*, ...)+249>
0x00000000003c6605 <???+3958277>
0x00000000003c67a2 <mmu::vm_fault(unsigned long, exception_frame*)+350>
0x000000000048bae5 <page_fault+310>
0x000000000048a976 <???+4761974>
0x000000000058db51 <???+5823313>
0x000000000058dc20 <???+5823520>
0x000000000058ec60 <free+147>
0x00000000006dd381 <???+7197569>
0xffffa00007378e7f <???+121081471>
0x0f1ff0ba0f660061 <???+258343009>

```

**Comment by benoit-canet** [comment]
Could you add subtree_check to your /etc/exports ?

**Comment by benoit-canet** [comment]
Are you sure to use latest git head ?

**Comment by justinc1** [comment]
Now, I'm not. switched branches up and down, and OSv v0.24-59-g26f1b98 - it doesn't tell me much about say last commit. Sufficient to say, on my PC I had wrong branch checked out (was looking at issue 713 yesterday), so on HLRS I had image built at - some other branch.

Benoit, did you had any problem with mounting subdirectory? I will repeat test with master@2b04baba + few commits

**Comment by benoit-canet** [comment]
I had issues but I was on the wrong branch.

With latest git head it work just fine.

**Comment by nyh** [comment]
Guys, I lost track of whether the latest problem you mentioned exists or doesn't exist because you used the wrong version of OSv. If a bug still exists, please open a new issue for it.

**Comment by masksasi** [comment]
HI all,

am getting the same issue when i tried to use ls on the NFS mounted directory in OSV.  could you please share me the working Git branch of OSv. 

Current setup:

Server exports:
============
/data	192.168.122.0/24(rw,async,no_subtree_check,no_root_squash,insecure,sec=sys)

client 
====
tried mount through API:

curl -X PUT http://192.168.122.89:8000/app?command=/tools/mount-nfs.so%20nfs://192.168.122.1/data/?uid=0%20/data


$ sudo ./scripts/run.py -V -v -n -p kvm 
OSv v0.24-197-g74543fc
4 CPUs detected
Firmware vendor: SeaBIOS
bsd: initializing - done
VFS: mounting ramfs at /
VFS: mounting devfs at /dev
net: initializing - done
vga: Add VGA device instance
eth0: ethernet address: 52:54:00:12:34:56
virtio-blk: Add blk device instances 0 as vblk0, devsize=10737418240
random: virtio-rng registered as a source.
random: intel drng, rdrand registered as a source.
random: <Software, Yarrow> initialized
VFS: unmounting /dev
VFS: mounting zfs at /zfs
zfs: mounting osv/zfs from device /dev/vblk0.1
random: device unblocked.
VFS: mounting devfs at /dev
VFS: mounting procfs at /proc
program zpool.so returned 1
BSD shrinker: event handler list found: 0xffffa00001e10c80
	BSD shrinker found: 1
BSD shrinker: unlocked, running
[I/42 dhcp]: Waiting for IP...
[I/244 dhcp]: Server acknowledged IP for interface eth0
eth0: 192.168.122.89
[I/244 dhcp]: Configuring eth0: ip 192.168.122.89 subnet mask 255.255.255.0 gateway 192.168.122.1 MTU 1500
Running from /init/30-auto-00: /libhttpserver.so &!
could not load libicui18n.so.55
could not load libicuuc.so.55
could not load libssl.so.1.0.0
could not load libcrypto.so.1.0.0
/# VFS: mounting nfs at /data

/# 
/# ls /
page fault outside application, addr: 0x0000000000000000
[registers]
RIP: 0x00000000004246f9 <vn_stat+89>
RFL: 0x0000000000010206  CS:  0x0000000000000008  SS:  0x0000000000000010
RAX: 0x0000000000000000  RBX: 0x00002000001fe7f0  RCX: 0x0000000000000000  RDX: 0x00000000000000b1
RSI: 0x00002000001fd450  RDI: 0xffffa00001be5100  RBP: 0x00002000001fd4e0  R8:  0x00002000001fb4b1
R9:  0x8080808080808001  R10: 0xffff80000493a040  R11: 0x726e6e616167ff2e  R12: 0xffffa00001be5100
R13: 0x8080808080808001  R14: 0xffff80000493a040  R15: 0x726e6e616167ff2e  RSP: 0x00002000001fd450
Aborted

[backtrace]
0x0000000000329ad8 <???+3316440>
0x000000000032aabb <mmu::vm_fault(unsigned long, exception_frame*)+171>
0x0000000000388be8 <page_fault+136>
0x0000000000387a86 <???+3701382>
0x0000000000426388 <sys_stat+56>
0x000000000041d26e <__xstat+78>
0x0000100000ca4824 <httpserver::api::file::get_file_handler::list_directory(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, http::server::request const&, http::server::reply&)+4276
